### PR TITLE
Added "How to Use a Custom Version Strategy for Assets"

### PR DESCRIPTION
**Pull request by javiereguiluz**
This PR finishes the great work made by @teohhanhui in #5489.

### CODE REVIEW

**Comment by stof** [comment]
you should also fix conflicts

File: frontend/custom_version_strategy.rst [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+applications by adding a version identifier to the URL of your static assets
+(CSS, JavaScript, images, etc.) When the content of the asset changes, the
+identifier changes and the browser is forced to download it again instead of
+using the cached version.
+
+Symfony supports the basic asset versioning thanks to the
+:ref:`version <reference-framework-assets-version>` and
+:ref:`version_format <reference-framework-assets-version-format>` configuration
+options. If your application requires a more advanced versioning, you can create
+your own version strategy.
```
Comment (OskarStark) [inline frontend/custom_version_strategy.rst]: Shouldn't we add an example here, what "requires a more advanced versioning"  means?

File: frontend/custom_version_strategy.rst [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        "css/style.css": "91cd067f79a5839536b46c494c4272d8"
+    }
+
+Implement VersionStrategyInterface
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Asset version strategies are PHP classes that implement the
+:class:`Symfony\\Component\\Asset\\VersionStrategy\\VersionStrategyInterface`.
+In this example, the constructor of the class takes as arguments the path to
+the manifest file generated by gulp-buster and the format of the generated
```
Comment (OskarStark) [inline frontend/custom_version_strategy.rst]: gulp buster should be a link here, too

File: reference/twig_reference.rst [hunk 3]
```diff
@@ -93,6 +93,8 @@ Returns an instance of ``ControllerReference`` to be used with functions
 like :ref:`render() <reference-twig-function-render>` and
 :ref:`render_esi() <reference-twig-function-render-esi>`.
 
+.. _reference-twig-function-asset:
```
Comment (OskarStark) [inline reference/twig_reference.rst]: why did you add this here @javiereguiluz ?

Comment (javiereguiluz) [inline reference/twig_reference.rst]: Good catch. It was in the original PR. I've just removed it.

File: frontend/custom_version_strategy.rst (hunk 2 above)
Comment (javiereguiluz) [inline frontend/custom_version_strategy.rst]: Done.

File: frontend/custom_version_strategy.rst (hunk 1 above)
Comment (javiereguiluz) [inline frontend/custom_version_strategy.rst]: I've reworded this part a bit. Thanks.

File: frontend/custom_version_strategy.rst [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+                '%%s?version=%%s',
+            )
+        );
+        $definition->setPublic(false);
+
+        $container->setDefinition('app.assets.versioning.gulp_buster', $definition);
+
+Finally, enable the new asset versioning for all the application assets or just
+for some :ref:`asset package <reference-framework-assets-packages>` thanks to
+the :ref:`version_strategy <reference-framework-assets-version_strategy>` option:
```
Comment (xabbuh) [inline frontend/custom_version_strategy.rst]: It looks like you have to add a label for this.

File: frontend/custom_version_strategy.rst [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+How to Use a Custom Version Strategy for Assets
+===============================================
+
+.. versionadded:: 2.7
+    The Asset component was introduced in Symfony 2.7.
+
+Asset versioning is a technique that improves the performance of web
+applications by adding a version identifier to the URL of the static assets
+(CSS, JavaScript, images, etc.) When the content of the asset changes, its
+identifier is also modified to force the browser download it again instead of
```
Comment (xabbuh) [inline frontend/custom_version_strategy.rst]: [...] the browser to download it [...]

File: frontend/custom_version_strategy.rst [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            if ($path && '/' === $path[0]) {
+                return '/'.$versionized;
+            }
+
+            return $versionized;
+        }
+
+        private function loadManifest(array $options)
+        {
+            $hashes = json_decode(file_get_contents($this->manifestPath), true);
```
Comment (xabbuh) [inline frontend/custom_version_strategy.rst]: The variable is not needed.

File: frontend/custom_version_strategy.rst [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+            return $hashes;
+        }
+    }
+
+Register the Strategy Service
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+After creating the strategy PHP class, register it as a Symfony service
```
Comment (xabbuh) [inline frontend/custom_version_strategy.rst]: Missing colon at the end of the sentence.

File: frontend/custom_version_strategy.rst [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+        <!-- app/config/config.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xmlns:framework="http://symfony.com/schema/dic/symfony"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd
+                http://symfony.com/schema/dic/symfony http://symfony.com/schema/dic/symfony/symfony-1.0.xsd">
+
+            <framework:config>
+                <framework:assets version_strategy="app.assets.versioning.gulp_buster" />
```
Comment (xabbuh) [inline frontend/custom_version_strategy.rst]: In XML, the attribute should be `version-strategy`.

File: reference/configuration/framework.rst [hunk 9]
```diff
@@ -1130,6 +1135,16 @@ is set to ``5``, the asset's path would be ``/images/logo.png?version=5``.
...
     asset versions to remain accessible at their original URL.
 
+.. _reference-framework-assets-version_strategy:
+
+version_strategy
+................
+
+**type**: ``string`` **default**: null
+
+The service id of the :doc:`asset version strategy </frontend/custom_version:strategy>`
```
Comment (xabbuh) [inline reference/configuration/framework.rst]: The colon in the path would be an underscore.

**Review by xabbuh — CHANGES_REQUESTED** [review]

**Comment by javiereguiluz** [comment]
@xabbuh thanks for your review. I've fixed everything.

File: reference/configuration/framework.rst [hunk 10]
```diff
@@ -1013,6 +1016,7 @@ Each package can configure the following options:
 * :ref:`base_urls <reference-assets-base-urls>`
 * :ref:`version <reference-framework-assets-version>`
 * :ref:`version_format <reference-assets-version-format>`
+* :ref:`version_strategy <reference-assets-version-strategy>`
```
Comment (xabbuh) [inline reference/configuration/framework.rst]: I guess this should be `reference-framework-assets-version_strategy` instead (see the build failure on Travis CI).

**Comment by xabbuh** [comment]
:+1:

Status: Reviewed

**Review by xabbuh — APPROVED** [review]

**Comment by xabbuh** [comment]
Thank you Javier.

**Comment by xabbuh** [comment]
And thank you for starting this awesome PR @teohhanhui!

**Pull request merged**